pkg/fqdn/dnsproxy/shared_client.go

```diff
@@ -50,29 +50,14 @@ func (s *SharedClients) ExchangeContext(ctx context.Context, key string, conf *d
 func (s *SharedClients) ShutdownClient(key string) {
 	// lock for s.clients access
```

Comment:
I wonder if it is worth checking the refcount here and print a log message if it's non-zero, since that would indicate a bug in our logic. 

On the other hand, it would probably spam the user logs, so I'm not sure it's worth it

Reply:
Yeah it's not a useful message to surface to the user, but I can make it a debug log?

Reply:
Thinking about this some more, we also invoke `ShutdownClient` if `TCPConn.Write` fails - which I think could happen with a non-zero refcount? In that case, we probably don't want to emit a log

Reply:
Yep, I've had to change approach and switch to letting the reference count increase since reference count zero is treated specially in the allocation logic and caused an infinite loop. I'm not emitting a log since there's not really a clear condition when to do so anymore now.